driver/resource: Support dediprog

edersondisouza opened the pull request:
Akin to 'flashrom', the Dediprog resource expects 'dpcmd' to be available on the host. Currently, only parameter that can be set is VCC.

<!---
Describe what your pull request does,
i.e. fix this bug and how, add a feature, fix documentation…
If you add a feature, please answer these questions:
- what do you use the feature for?
- how does labgrid benefit as a testing library from the feature?
- how did you verify the feature works?
- if hardware is needed for the feature, which hardware is supported and which
  hardware did you test with?
--->
**Description**
The main advantage to use it instead of `flashrom` is that it allows for using images smaller than the SPI memory, which is then automatically filled with `0xff`. For `flashrom`, one needs the image to have the right size - although one can avoid writing the whole of it by having a layout file - which adds to the inconvenience. Using `dpcmd` is more straightforward and faster.
Tested locally using an SF100.

<!---
This checklist roughly outlines the steps for new features, remove and add tasks as needed:
--->
**Checklist**
- [X] Documentation for the feature
- [X] Tests for the feature 
<!---
If you add a driver/resource or modify one:
--->
- [X] The arguments and description in doc/configuration.rst have been updated

edersondisouza commented:
v2:
 - Fixed documentation build issues.

Emantor commented on doc/configuration.rst:
```diff
@@ -1025,6 +1025,27 @@ resource available on a remote computer
+.. code-block:: yaml
+
+  tools:
+    dpcmd: '/usr/sbin/dpcmd'
```
This tools entry belongs to the `DediprogDriver`, since it uses the tool. The example here should be something like:

```
Dediprog:
  vcc: 0
```

Also maybe we should rename this from `Dediprog` to `DediprogFlasher` and the Driver to `DediprogFlashDriver` since dediprog also builds eMMC programmers.

edersondisouza replied:
While I did add a `vcc` example usage, I think that the tool really belongs here. This is similar to `Flashrom`, as where the tool is located can vary among exporters, so it's a resource attribute. Or did I understand `Flashrom` case wrong?

Emantor replied:
You're right. I'm now wondering how we handle this on the exporter side, but thats not relevant for this PR.

jluebbe commented on labgrid/driver/dediprogdriver.py:
```diff
@@ -0,0 +1,67 @@
+        arg_list = list(args)
+        arg_list.append('--vcc')
+        arg_list.append(f'{self.dediprog_resource.vcc}')
+        arg_list.append('--silent')
```
Is there still enough output to debug problems even with `--silent`?

edersondisouza replied:
Yep. The `--silent` just prevents a "fancy" time counter.

jluebbe commented on doc/configuration.rst:
```diff
@@ -1025,6 +1025,27 @@ resource available on a remote computer
+    dpcmd: '/usr/sbin/dpcmd'
+
+Arguments:
+  - vcc (int): 0 (3.5V), 1 (2.5V), 2 (1.8V)
```
Hmm, I'd prefer to have the actual voltage in the configuration file, as this makes it notice a wrong configuration. You can map the voltages to the command line arguments in the driver.

edersondisouza replied:
Fair.

edersondisouza commented:
v3:
 - Name resources `DediprogFlasher` and `NetworkDediprogFlasher`, and driver `DediprogFlashDriver`;
 - Use 'named' vcc argument: `3.5V` instead of `0`, for instance.

edersondisouza commented:
v4:
- Add some more tests that will help increase coverage.

edersondisouza commented:
Any comments? @Emantor @jluebbe

Bastian-Krause commented on labgrid/driver/dediprogflashdriver.py:
```diff
@@ -0,0 +1,77 @@
+
+@target_factory.reg_driver
+@attr.s(eq=False)
+class DediprogFlashDriver(Driver, BootstrapProtocol):
```
@jluebbe @Emantor I always thought that the BootstrapProtocol was meant to write to non-persistent memory only (i.e. load an image into RAM). But the FlashromDriver seems to also implement it and now DediprogFlashDriver will, too?

In #206 we discussed that the QuartusHPSDriver explicitly does not implement the BootstrapProtocol. The plan was to introduce some kind of FlashProtocol instead later.

edersondisouza replied:
Ok, will drop `BootstrapProtocol` on my next version. And don't really need `labgrid-client` support, as I'm using python library directly.

edersondisouza commented:
v5:
 - Dropped support for `BootstrapProtocol`, exposing `flash()` method;
 - Added `erase()` method;
 - Rebased.

Bastian-Krause commented on labgrid/resource/dediprogflasher.py:
```diff
@@ -0,0 +1,16 @@
+@attr.s(eq=False)
+class DediprogFlasher(Resource):
+    """vcc: '3.5V', '2.5V' or '1.8V'"""
+    vcc = attr.ib(validator=attr.validators.instance_of(str))
```
```suggestion
    vcc = attr.ib(validator=attr.validators.in_(('3.5V', '2.5V', '1.8V')))
```
See [attrs.validators.in_()](https://www.attrs.org/en/stable/api.html#attrs.validators.in_).

Bastian-Krause commented on labgrid/resource/dediprogflasher.py:
```diff
@@ -0,0 +1,16 @@
+@target_factory.reg_resource
+@attr.s(eq=False)
+class DediprogFlasher(Resource):
+    """vcc: '3.5V', '2.5V' or '1.8V'"""
```
We can drop that if we use the suggestion below.

Bastian-Krause commented on labgrid/driver/dediprogflashdriver.py:
```diff
@@ -0,0 +1,78 @@
+    def on_deactivate(self):
+        pass
+
+    def map_vcc(self):
+        vcc_map = {'3.5V': '0', '2.5V': '1', '1.8V': '2'}
+        vcc = self.dediprog_resource.vcc
+        if vcc not in vcc_map:
+            raise ValueError(f"VCC {vcc} not '1.8V', '2.5V' or '3.5V'")
+
+        return vcc_map[vcc]
```
I guess it's better to check that in the resource class, suggestion below.

Bastian-Krause commented on labgrid/resource/dediprogflasher.py:
```diff
@@ -0,0 +1,16 @@
+@target_factory.reg_resource
+@attr.s(eq=False)
+class NetworkDediprogFlasher(NetworkResource):
+    vcc = attr.ib(validator=attr.validators.instance_of(str))
```
Same here.

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,43 @@
+    d = DediprogFlashDriver(target, name=None)
+    assert (isinstance(d, DediprogFlashDriver))
+
+def test_dediprog_driver_map_vcc(target):
+    r = DediprogFlasher(target,
+                 name=None,
+                 vcc='2.5V')
+    d = DediprogFlashDriver(target, name=None)
+
+    m = d.map_vcc()
+    assert (isinstance(m, str))
+    assert (m == '1')
+
+def test_dediprog_driver_map_vcc_inv(target):
+    r = DediprogFlasher(target,
+                 name=None,
+                 vcc='2.6V')
+    d = DediprogFlashDriver(target, name=None)
+
+    with pytest.raises(ValueError):
+        m = d.map_vcc()
```
We can drop this if we use the suggestion above.

Bastian-Krause replied:
I confused the check for valid values and the mapping here.

edersondisouza commented:
v6:
 - Using attr `in_` to map validate vcc argument;
 - Added `dpcmd` entry to `labgrid-device-config` doc;
 - Rebased.

Bastian-Krause commented on labgrid/driver/dediprogflashdriver.py:
```diff
@@ -0,0 +1,72 @@
+        'dediprog_resource': {"DediprogFlasher", NetworkDediprogFlasher},
+    }
+
+    image = attr.ib(default=None)
```
```suggestion
    image = attr.ib(validator=attr.validators.instance_of(str), default=None)
```
The docs say that we expect a string here, so let's check for that.

Bastian-Krause commented on labgrid/driver/dediprogflashdriver.py:
```diff
@@ -0,0 +1,72 @@
+        self.logger.debug('Tool %s', self.tool)
+
+    def _get_dediprog_prefix(self):
+        return self.dediprog_resource.command_prefix+[
+            self.tool
+        ]
```
```suggestion
        return self.dediprog_resource.command_prefix + [self.tool]
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+                       host='localhost')
+    assert (isinstance(r, NetworkDediprogFlasher))
+    d = DediprogFlashDriver(target, name=None)
+    assert (isinstance(d, DediprogFlashDriver))
```
```suggestion
    assert isinstance(d, DediprogFlashDriver)
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+                       name=None,
+                       vcc='3.5V',
+                       host='localhost')
+    assert (isinstance(r, NetworkDediprogFlasher))
```
```suggestion
    assert isinstance(r, NetworkDediprogFlasher)
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+    r = DediprogFlasher(target,
+                 name=None,
+                 vcc='2.5V')
+    assert (isinstance(r, DediprogFlasher))
```
```suggestion
    assert isinstance(r, DediprogFlasher)
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+    assert (isinstance(r, DediprogFlasher))
+
+    d = DediprogFlashDriver(target, name=None)
+    assert (isinstance(d, DediprogFlashDriver))
```
```suggestion
    assert isinstance(d, DediprogFlashDriver)
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+    d = DediprogFlashDriver(target, name=None)
+
+    m = d.map_vcc()
+    assert (isinstance(m, str))
```
```suggestion
    assert isinstance(m, str)
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+
+    m = d.map_vcc()
+    assert (isinstance(m, str))
+    assert (m == '1')
```
```suggestion
    assert m == '1'
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+
+
+def test_dediprog_network_create(target):
+    r = NetworkDediprogFlasher(target,
+                       name=None,
+                       vcc='3.5V',
+                       host='localhost')
```
```suggestion
    r = NetworkDediprogFlasher(target, name=None, vcc='3.5V', host='localhost')
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+
+
+def test_dediprog_driver_create(target):
+    r = DediprogFlasher(target,
+                 name=None,
+                 vcc='2.5V')
```
```suggestion
    r = DediprogFlasher(target, name=None, vcc='2.5V')
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+    assert (isinstance(d, DediprogFlashDriver))
+
+def test_dediprog_driver_map_vcc(target):
+    r = DediprogFlasher(target,
+                 name=None,
+                 vcc='2.5V')
```
```suggestion
    r = DediprogFlasher(target, name=None, vcc='2.5V')
```

Bastian-Krause commented on tests/test_dediprogflasher.py:
```diff
@@ -0,0 +1,40 @@
+
+def test_dediprog_driver_map_vcc_inv(target):
+    with pytest.raises(ValueError):
+        r = DediprogFlasher(target,
+                     name=None,
+                     vcc='2.6V')
```
```suggestion
        r = DediprogFlasher(target, name=None, vcc='2.6V')
```

Bastian-Krause commented on labgrid/driver/dediprogflashdriver.py:
```diff
@@ -0,0 +1,72 @@
+            self.tool = self.target.env.config.get_tool('dpcmd')
+        else:
+            self.tool = 'dpcmd'
+        self.logger.debug('Tool %s', self.tool)
```
```suggestion
```
I don't think this is useful, `processwrapper.check_output()` logs the command anyway and passes it to the StepReporter.

Bastian-Krause commented on labgrid/driver/dediprogflashdriver.py:
```diff
@@ -0,0 +1,72 @@
+    """ The DediprogFlashDriver uses the dediprog utility to write an image
+    to a raw rom. The driver is a pure wrapper of the dpcmd utility"""
+    bindings = {
+        'dediprog_resource': {"DediprogFlasher", NetworkDediprogFlasher},
```
Hm, I'm not too happy about the name of the binding. Maybe `flasher` is better?

Bastian-Krause reviewed (changes requested):
Thanks for working on this. I've added some more comments.

edersondisouza commented:
v7:
  - Fixed styling issues;
  - Ensure `image` attr is actually a string;
  - Removed redundant log message;
  - Renamed resource to  `flasher`;
  - Rebased.

jluebbe commented:
Note that there is no support for udev-matching and so no support in the exporter. The `NetworkDediprogFlasher` needs to be configured manually in the local environments.

Bastian-Krause commented on labgrid/driver/dediprogflashdriver.py:
```diff
@@ -0,0 +1,69 @@
+        'flasher': {"DediprogFlasher", NetworkDediprogFlasher},
+    }
+
+    image = attr.ib(validator=attr.validators.instance_of(str), default='')
```
```suggestion
    image = attr.ib(validator=attr.validators.optional(attr.validators.instance_of(str)), default=None)
```
Oh, I'm sorry that my suggestion did not work. That's what I meant, actually.

Bastian-Krause replied:
Fixed it myself, so we can get this merged.

Bastian-Krause reviewed (changes requested):
Everything else looks good.
